File: 219-contains-duplicate-ii/contains-duplicate-ii.py
class Solution:
    def containsNearbyDuplicate(self, nums: List[int], k: int) -> bool:
        hm = {}

        i = 0
        j = 1
        while j < len(nums):
            if nums[i] not in hm:
                hm[nums[i]] = i

            if nums[j] not in hm:
                i = j
            else:
                if abs(j - hm[nums[j]]) <= k:
                    return True
                else:
                    hm[nums[j]] = j
            j+=1
        return False

        # hs = set()

        # i = 0
        # j = 1
        # while j<len(nums):
        #     if nums[i] not in hs:
        #         hs.add(nums[i])
            

File: 219-contains-duplicate-ii/contains-duplicate-ii.java
class Solution {
    public boolean containsNearbyDuplicate(int[] nums, int k) {

        HashMap<Integer, Integer> hm = new HashMap<>();
        for(int i=0;i<=k && i < nums.length;i++){
            if(!hm.containsKey(nums[i])){
                hm.put(nums[i], i);
            }
            else{
                if(Math.abs(i-hm.get(nums[i]))<=k){
                    return true;
                }
                else{
                    hm.put(nums[i], i);
                }
            }
        }
        System.out.println(hm);


        for(int j=k+1;j<nums.length;j++){
            hm.remove(nums[j-k-1]);
            if(hm.containsKey(nums[j]) && Math.abs(j-hm.get(nums[j]))<=k ){
                return true;
            }
            else{
                hm.put(nums[j], j);
            }
        }

        return false;


    }
}